docs/Impredicativity.md

```diff
@@ -0,0 +1,142 @@
+# `Lens` vs `ALens`: a Note on Impredicativity
+
+Sometimes, you may encounter some strange errors to do with types that
+_should_ unify but don't. For example:
+
+```purescript
+newtype Wrapped = Wrapped String
+derive instance newtypeWrapped :: Newtype Wrapped _
+
+-- `_Newtype` is an `Iso'`, so we can specialise to `Lens'`.
+good :: Lens' Wrapped String
+good = _Newtype
+
+-- So, we should be able to do the same when lifted into a record.
+type Example = { lens :: Iso' Wrapped String }
+
+bad :: Example -> Lens' Wrapped String
+bad = _.lens
+
+-- Could not match constrained type
+--    Profunctor t2 => t2 String String -> t2 Wrapped Wrapped
+--  with type
+--    p0 String String -> p0 Wrapped Wrapped
+```
+
+Below, we'll discuss how these problems arise, and what we can do to overcome
+them.
+
+## The Long Answer
+
+The subtle issue here relates to a subject called *impredicativity*. We can make
+a smaller reproducible example:
+
+```purescript
+{ foo: id } :: { foo :: forall a. a -> a }
+```
+
+We can, as we would expect, take `foo` out and cast it to whatever we like:
+
+```purescript
+({ foo: id } :: { foo :: forall a. a -> a }).foo
+  :: String -> String
+```
+
+So, while the original type said that we could pick _any_ `a`, we restricted that
+to `String` afterwards. In some situations, such as the above, the type-checker
+will cope happily. However, the troubles usually arrive when you extract a value
+with **a constraint**. Let's generalise the `id` function a little further:
+
+```purescript
+({ foo: id } :: { foo :: forall k a. Category k => k a a }).foo
+ :: String -> String
+```
+
+Now, we've generalised our `foo` to work with any `Category`, and not just
+`(->)`. However, we've also got ourselves a type error! _What did we do?_
+
+Let's break this line down into two parts:
+
+1. We extract `forall k a. foo :: Category k => k a a` out our record.
```

Comment:
This is not valid syntax

Reply:
iirc I think this was deliberate to try to illustrate scoping, but has clearly caused more confusion than it saved :sweat_smile: